"AssertionError: stmt_val !== nothing" in compiler

The following error occurs in `v0.7.-beta`, but did not occur up to and including `v0.7.0-alpha`.

`git bisect` points to commit a0de85d866557f5d4d505c232b1aa3f4053cddfb.

```jl
ERROR: LoadError: AssertionError: stmt_val !== nothing
Stacktrace:
 [1] perform_lifting!(::Core.Compiler.IncrementalCompact, ::Array{Any,1}, ::Any, ::Core.Compiler.IdDict{Core.Compiler.Pair{Union{Core.Compiler.NewSSAValue, Core.Compiler.OldSSAValue, Core.SSAValue},Any},Union{Core.Compiler.NewSSAVa
lue, Core.Compiler.OldSSAValue, Core.SSAValue}}, ::Any, ::Core.Compiler.IdDict{Any,Any}, ::Any) at ./compiler/ssair/passes.jl:483
 [2] getfield_elim_pass!(::Core.Compiler.IRCode, ::Core.Compiler.DomTree) at ./compiler/ssair/passes.jl:660
 [3] run_passes(::Core.CodeInfo, ::Int64, ::Core.Compiler.OptimizationState) at ./compiler/ssair/driver.jl:123
 [4] optimize(::Core.Compiler.OptimizationState, ::Any) at ./compiler/optimize.jl:146
 [5] typeinf(::Core.Compiler.InferenceState) at ./compiler/typeinfer.jl:35
 [6] typeinf_edge(::Method, ::Any, ::Core.SimpleVector, ::Core.Compiler.InferenceState) at ./compiler/typeinfer.jl:544
 [7] abstract_call_method(::Method, ::Any, ::Core.SimpleVector, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:303
 [8] abstract_call_gf_by_type(::Any, ::Array{Any,1}, ::Any, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:78
 [9] abstract_call(::Any, ::Array{Any,1}, ::Array{Any,1}, ::Array{Any,1}, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:709
 [10] abstract_eval_call(::Array{Any,1}, ::Array{Any,1}, ::Array{Any,1}, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:738
 [11] abstract_eval(::Any, ::Array{Any,1}, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:825
 [12] typeinf_local(::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:1035
 [13] typeinf_nocycle(::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:1105
 [14] typeinf(::Core.Compiler.InferenceState) at ./compiler/typeinfer.jl:15
 [15] typeinf_edge(::Method, ::Any, ::Core.SimpleVector, ::Core.Compiler.InferenceState) at ./compiler/typeinfer.jl:544
 [16] abstract_call_method(::Method, ::Any, ::Core.SimpleVector, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:303
 [17] abstract_call_gf_by_type(::Any, ::Array{Any,1}, ::Any, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:78
 [18] abstract_call(::Any, ::Array{Any,1}, ::Array{Any,1}, ::Array{Any,1}, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:709
 [19] abstract_eval_call(::Array{Any,1}, ::Array{Any,1}, ::Array{Any,1}, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:738
 [20] abstract_eval(::Any, ::Array{Any,1}, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:825
 [21] typeinf_local(::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:1049
 [22] typeinf_nocycle(::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:1105
 [23] typeinf(::Core.Compiler.InferenceState) at ./compiler/typeinfer.jl:15
 [24] typeinf_edge(::Method, ::Any, ::Core.SimpleVector, ::Core.Compiler.InferenceState) at ./compiler/typeinfer.jl:544
 [25] abstract_call_method(::Method, ::Any, ::Core.SimpleVector, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:303
 [26] abstract_call_gf_by_type(::Any, ::Array{Any,1}, ::Any, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:78
 [27] abstract_call(::Any, ::Array{Any,1}, ::Array{Any,1}, ::Array{Any,1}, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:709
 [28] abstract_eval_call(::Array{Any,1}, ::Array{Any,1}, ::Array{Any,1}, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:738
 [29] abstract_eval(::Any, ::Array{Any,1}, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:825
 [30] typeinf_local(::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:1049
 [31] typeinf_nocycle(::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:1105
 [32] typeinf(::Core.Compiler.InferenceState) at ./compiler/typeinfer.jl:15
 [33] typeinf_edge(::Method, ::Any, ::Core.SimpleVector, ::Core.Compiler.InferenceState) at ./compiler/typeinfer.jl:544
 [34] abstract_call_method(::Method, ::Any, ::Core.SimpleVector, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:303
 [35] abstract_call_gf_by_type(::Any, ::Array{Any,1}, ::Any, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:78
 [36] abstract_call(::Any, ::Array{Any,1}, ::Array{Any,1}, ::Array{Any,1}, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:709
 [37] abstract_eval_call(::Array{Any,1}, ::Array{Any,1}, ::Array{Any,1}, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:738
 [38] abstract_eval(::Any, ::Array{Any,1}, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:825
 [39] typeinf_local(::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:1049
 [40] typeinf_nocycle(::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:1105
 [41] typeinf(::Core.Compiler.InferenceState) at ./compiler/typeinfer.jl:15
 [42] typeinf_edge(::Method, ::Any, ::Core.SimpleVector, ::Core.Compiler.InferenceState) at ./compiler/typeinfer.jl:544
 [43] abstract_call_method(::Method, ::Any, ::Core.SimpleVector, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:303
 [44] abstract_call_gf_by_type(::Any, ::Array{Any,1}, ::Any, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:78
 [45] abstract_call(::Any, ::Array{Any,1}, ::Array{Any,1}, ::Array{Any,1}, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:709
 [46] abstract_eval_call(::Array{Any,1}, ::Array{Any,1}, ::Array{Any,1}, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:738
 [47] abstract_eval(::Any, ::Array{Any,1}, ::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:825
 [48] typeinf_local(::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:1049
 [49] typeinf_nocycle(::Core.Compiler.InferenceState) at ./compiler/abstractinterpretation.jl:1105
 [50] typeinf(::Core.Compiler.InferenceState) at ./compiler/typeinfer.jl:15
 [51] typeinf_edge(::Method, ::Any, ::Core.SimpleVector, ::Core.Compiler.InferenceState) at ./compiler/typeinfer.jl:544
```

--- a/base/compiler/ssair/passes.jl
+++ b/base/compiler/ssair/passes.jl
@@ -399,12 +399,13 @@ function lift_comparison!(compact::IncrementalCompact, idx::Int,
     end
 
     lifted_val = perform_lifting!(compact, visited_phinodes, cmp, lifting_cache, Bool, lifted_leaves, val)
+    @assert lifted_val !== nothing
 
     #global assertion_counter
     #assertion_counter::Int += 1
     #insert_node_here!(compact, Expr(:assert_egal, Symbol(string("assert_egal_", assertion_counter)), SSAValue(idx), lifted_val), nothing, 0, true)
     #return
-    compact[idx] = lifted_val
+    compact[idx] = lifted_val.x
 end
 
 struct LiftedPhi
@@ -478,13 +479,12 @@ function perform_lifting!(compact::IncrementalCompact,
     if isa(stmt_val, Union{SSAValue, OldSSAValue})
         stmt_val = simple_walk(compact, stmt_val)
     end
+
     if stmt_val in keys(lifted_leaves)
         stmt_val = lifted_leaves[stmt_val]
-        @assert stmt_val !== nothing
-        stmt_val = stmt_val.x
     else
         isa(stmt_val, Union{SSAValue, OldSSAValue}) && stmt_val in keys(reverse_mapping)
-        stmt_val = lifted_phis[reverse_mapping[stmt_val]].ssa
+        stmt_val = RefValue{Any}(lifted_phis[reverse_mapping[stmt_val]].ssa)
     end
 
     return stmt_val
@@ -661,14 +661,20 @@ function getfield_elim_pass!(ir::IRCode, domtree)
 
         # Insert the undef check if necessary
         if any_undef && !is_unchecked
-            insert_node!(compact, SSAValue(idx), Nothing, Expr(:undefcheck, :getfield, val))
+            if val === nothing
+                insert_node!(compact, SSAValue(idx), Nothing, Expr(:throw_undef_if_not, :getfield, false))
+            else
+                insert_node!(compact, SSAValue(idx), Nothing, Expr(:undefcheck, :getfield, val.x))
+            end
+        else
+            @assert val !== nothing
         end
 
         global assertion_counter
         assertion_counter::Int += 1
         #insert_node_here!(compact, Expr(:assert_egal, Symbol(string("assert_egal_", assertion_counter)), SSAValue(idx), val), nothing, 0, true)
         #continue
-        compact[idx] = val
+        compact[idx] = val === nothing ? nothing : val.x
     end
 
     ir = finish(compact)

--- a/base/compiler/ssair/passes.jl
+++ b/base/compiler/ssair/passes.jl
@@ -662,9 +662,9 @@ function getfield_elim_pass!(ir::IRCode, domtree)
         # Insert the undef check if necessary
         if any_undef && !is_unchecked
             if val === nothing
-                insert_node!(compact, SSAValue(idx), Nothing, Expr(:throw_undef_if_not, :getfield, false))
+                insert_node!(compact, SSAValue(idx), Nothing, Expr(:throw_undef_if_not, Symbol("##getfield##"), false))
             else
-                insert_node!(compact, SSAValue(idx), Nothing, Expr(:undefcheck, :getfield, val.x))
+                insert_node!(compact, SSAValue(idx), Nothing, Expr(:undefcheck, Symbol("##getfield##"), val.x))
             end
         else
             @assert val !== nothing

--- a/src/ast.c
+++ b/src/ast.c
@@ -67,7 +67,7 @@ jl_sym_t *macrocall_sym; jl_sym_t *colon_sym;
 jl_sym_t *hygienicscope_sym;
 jl_sym_t *escape_sym;
 jl_sym_t *gc_preserve_begin_sym; jl_sym_t *gc_preserve_end_sym;
-jl_sym_t *throw_undef_if_not_sym;
+jl_sym_t *throw_undef_if_not_sym; jl_sym_t *getfield_undefref_sym;
 
 static uint8_t flisp_system_image[] = {
 #include <julia_flisp.boot.inc>
@@ -392,6 +392,7 @@ void jl_init_frontend(void)
     generated_sym = jl_symbol("generated");
     generated_only_sym = jl_symbol("generated_only");
     throw_undef_if_not_sym = jl_symbol("throw_undef_if_not");
+    getfield_undefref_sym = jl_symbol("##getfield##");
     do_sym = jl_symbol("do");
 }
 

--- a/src/codegen.cpp
+++ b/src/codegen.cpp
@@ -3928,8 +3928,14 @@ static jl_cgval_t emit_expr(jl_codectx_t &ctx, jl_value_t *expr, ssize_t ssaval)
         return emit_isdefined(ctx, args[0]);
     }
     else if (head == throw_undef_if_not_sym) {
-        Value *cond = emit_unbox(ctx, T_int8, emit_expr(ctx, args[1]), (jl_value_t*)jl_bool_type);
-        undef_var_error_ifnot(ctx, ctx.builder.CreateTrunc(cond, T_int1), (jl_sym_t*)args[0]);
+        jl_sym_t *var = (jl_sym_t*)args[0];
+        Value *cond = ctx.builder.CreateTrunc(emit_unbox(ctx, T_int8, emit_expr(ctx, args[1]), (jl_value_t*)jl_bool_type), T_int1);
+        if (var == getfield_undefref_sym) {
+            raise_exception_unless(ctx, cond,
+                literal_pointer_val(ctx, jl_undefref_exception));
+        } else {
+            undef_var_error_ifnot(ctx, cond, var);
+        }
         return ghostValue(jl_void_type);
     }
     else if (head == invoke_sym) {

--- a/src/interpreter.c
+++ b/src/interpreter.c
@@ -466,7 +466,11 @@ SECT_INTERP static jl_value_t *eval_value(jl_value_t *e, interpreter_state *s)
         jl_value_t *cond = eval_value(args[1], s);
         assert(jl_is_bool(cond));
         if (cond == jl_false) {
-            jl_undefined_var_error((jl_sym_t*)args[0]);
+            jl_sym_t *var = (jl_sym_t*)args[0];
+            if (var == getfield_undefref_sym)
+                jl_throw(jl_undefref_exception);
+            else
+                jl_undefined_var_error(var);
         }
         return jl_nothing;
     }

--- a/src/julia_internal.h
+++ b/src/julia_internal.h
@@ -973,6 +973,7 @@ extern jl_sym_t *gc_preserve_begin_sym; extern jl_sym_t *gc_preserve_end_sym;
 extern jl_sym_t *generated_sym;
 extern jl_sym_t *generated_only_sym;
 extern jl_sym_t *throw_undef_if_not_sym;
+extern jl_sym_t *getfield_undefref_sym;
 
 struct _jl_sysimg_fptrs_t;
 

--- a/test/core.jl
+++ b/test/core.jl
@@ -6267,3 +6267,6 @@ get27770(n::Nullable27770, v::Handle27770) = n.hasvalue ? n.value : v
 
 foo27770() = get27770(Nullable27770(), Handle27770())
 @test foo27770().ptr == Ptr{Cvoid}(UInt(0xfeedface))
+
+bar27770() = Nullable27770().value
+@test_throws UndefRefError bar27770()

--- a/test/core.jl
+++ b/test/core.jl
@@ -6250,3 +6250,20 @@ end
 wrap22291(ind) = (ind...,)
 @test @inferred(wrap22291(1)) == (1,)
 @test @inferred(wrap22291((1, 2))) == (1, 2)
+
+# Issue 27770
+mutable struct Handle27770
+    ptr::Ptr{Cvoid}
+end
+Handle27770() = Handle27770(Ptr{Cvoid}(UInt(0xfeedface)))
+
+struct Nullable27770
+    hasvalue::Bool
+    value::Handle27770
+    Nullable27770() = new(false)
+    Nullable27770(v::Handle27770) = new(true, Handle27770)
+end
+get27770(n::Nullable27770, v::Handle27770) = n.hasvalue ? n.value : v
+
+foo27770() = get27770(Nullable27770(), Handle27770())
+@test foo27770().ptr == Ptr{Cvoid}(UInt(0xfeedface))

